File: DailyChallenges/CPP/LeetCode_491.cpp
class Solution {
public:
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        int n = nums.size();
        set<vector<int>> s;
        function<void(int, vector<int>&)> dfs = [&] (int i, vector<int>& v) {
            if (i>=n) return;
            if (nums[i] >= v.back()){
                v.push_back(nums[i]);
                s.insert(v);
                dfs(i+1, v);
                v.pop_back();
            }
            dfs(i+1, v);
        };
        for (int i=0; i<n-1; ++i){
            vector<int> v; v.push_back(nums[i]);
            dfs(i+1, v);  
        }
        vector<vector<int>> ans(s.begin(), s.end());
        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_491.scala
object Solution {
    def findSubsequences(nums: Array[Int]): List[List[Int]] = {
        var ans = List.empty[List[Int]]
        
        def dfs(idx: Int, sub: List[Int]): Unit = {
            if (sub.length > 1) ans = sub.reverse :: ans
            val mark = scala.collection.mutable.HashSet.empty[Int]
            for (i <- idx until nums.length; if (sub.isEmpty || nums(i) >= sub.head) && !mark.contains(nums(i))){
                dfs(i+1, nums(i) :: sub)
                mark.add(nums(i))
            }
        }
        dfs(0, List.empty[Int])

        ans
    }
}
